### DISCUSSION TITLE
Memory error when working in Docker

### DISCUSSION THREAD

**Discussion by grigoryaleksanyan** [eclipse-birt/birt#849] [Q&A] [answered]
Hi!
I deployed WebViewer on a Tomcat server inside Docker.
Everything is fine, the image is created, the container started, Birt is running and I have access to it.

I'm generating an xlsx report, via the URL: http://localhost:8080/birt-viewer/run?__format=xlsx&__report=test.rptdesign&sample=my+parameter

Everything is fine, I can run the report and get the result.

This container is used as a report execution service. It receives requests from RabbitMQ queue for execution and has to generate and return a result file. Everything works fine until an error occurs: java.lang.OutOfMemoryError: Java heap space.

The bottom line is that the memory is not being freed after the report runs. And when ordering various large reports, the memory will run out.

I've tried increasing the size of available memory through the -Xmx8g -Xms4g -XX:MaxPermSize=1024m parmeters, but this only increases the uptime, and doesn't solve the problem in any way. I still get an OutOfMemoryError and my Tomcat server stops responding until I restart the container.

Can you please tell me how I can get Birt to clear the memory after the report is generated?

Sorry for my English :)

Birt 4.8

My Dockerfile: 
FROM tomcat:9.0
ENV JAVA_OPTS="-Xmx8g -Xms4g -XX:MaxPermSize=1024m"
COPY ["BirtViewer", "/usr/local/tomcat/webapps/birt-viewer"]
EXPOSE 8080
CMD ["catalina.sh", "run"]

**Comment by hvbtup** [comment]
Are you sure that the memory isn't freed?
You didn't mention this explicitly, but are you running the reports strictly sequentially or in parallel?
If you are running several reports at the same time, the amount of memory needed sums up of course.
You have to assure that you are not running too many reports at the same time.

**Reply by grigoryaleksanyan** [reply]
Yes, I'm sure.

For example, when sending http://localhost:8080/birt-viewer/run?__format=xlsx&__report=test.rptdesign&sample=my+parameter request to Birt, the amount of used memory increases by 1mb, after execution of report and getting xlsx file this 1mb is not freed. Accordingly, each such request will increase the memory, which eventually leads to an overflow and the following error: java.lang.OutOfMemoryError: Java heap space.
At full load, I send 16 reports for parallel execution. But even if you just request execution of one report, the memory will be allocated, but it won't free up after execution.

If it helps, here are the logs for the day when the error happened. 
https://drive.google.com/file/d/1hKVILY72W8-34Tk2DScQ0v20XYml0Ymq/view?usp=sharing

**Reply by hvbtup** [reply, marked as answer]
Just to be sure: How do you check the amount of memory used?

Note that Java will normally *not* give back memory to the OS even if it is no longer needed after a GC.
The JVM args -XX:MaxHeapFreeRatio and -XX:MinHeapFreeRatio can help.

But your message looks like it is indeed a problem of not enough heap memory in the JVM.
I'm not using the viewer app (instead I'm using a self-made "BIRT server app"), but I can hardly imagine this is a BIRT problem.
Maybe it is the data that you supply to BIRT which is not freed?
Or the data is cached (well, BIRT uses caching AFAIK, but I'm not enough of an expert here)?

**Reply by grigoryaleksanyan** [reply]
Ok! Thank you for your answer.
I check the memory inside the container using the "top" command:
https://drive.google.com/file/d/12rCpUJwZWAgJId7zN60kQGRvm8L9LkQ1/view?usp=sharing

**Reply by grigoryaleksanyan** [reply]
We replaced Tomcat image with Java 11 version to Java 8 version. We reduced the number of simultaneously processed reports from 16 to 5.

At the moment we could not get the error java.lang.OutOfMemoryError: Java heap space.

We will continue to watch for Birt.

**Comment by claesrosell** [comment]
@grigoryaleksanyan : Can you test this with a recent 4.9 build? If the problem is reproducible in 4.9 I can take a look at it.
Also, I would need a .rptdesign file, without dependencies to external datasource with which the problem is reproducible.

Another alternative  is that you provide a HeapDump file after the error has happend. This could be done by adding:
-XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=<file-or-dir-path>
as a JVM option. The HeapDump file will be named : have a .hprof file extension

**Reply by grigoryaleksanyan** [reply]
Ok! Thank you for your answer.
I will write to you in the near future

**Reply by grigoryaleksanyan** [reply]
Unfortunately it is not possible to test with version 4.9.  When running any report, Tomcat gives an error: 

HTTP Status 404 - Not found
Type: Status Report
Description: The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.
